enterprise/coordinator/recovery/recovery_test.go

```diff
@@ -0,0 +1,272 @@
+			t.Fatalf("no object received within %s", d.String())
+		case v := <-ch:
+			return v
+		}
+	}
+}
+
+func TestRecoverOnce(t *testing.T) {
+	logger := slog.Default()
+	ctx := t.Context()
+
+	for name, tc := range map[string]struct {
+		peerGetter peerGetter
+		guard      guard
+		dial       func(_ context.Context, issuer atls.Issuer, validators []atls.Validator, _ *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error)
+		wantErr    error
+	}{
+		"no peers": {
+			peerGetter: &fakePeerGetter{nil, nil},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return newFakeClient(t), func() error { return nil }, nil
+			},
+			wantErr: errNoPeers,
+		},
+		"bad peerGetter": {
+			peerGetter: &fakePeerGetter{nil, assert.AnError},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return newFakeClient(t), func() error { return nil }, nil
+			},
+			wantErr: assert.AnError,
+		},
+		"bad dial": {
+			peerGetter: &fakePeerGetter{[]string{"foo"}, nil},
+			guard:      newStaleGuard(t),
+			dial: func(context.Context, atls.Issuer, []atls.Validator, *slog.Logger, string) (meshapi.MeshAPIClient, func() error, error) {
+				return nil, nil, assert.AnError
+			},
+			wantErr: assert.AnError,
+		},
+		"one bad peer": {
+			peerGetter: &fakePeerGetter{[]string{"a", "b"}, nil},
+			guard:      newStaleGuard(t),
+			dial: func(_ context.Context, _ atls.Issuer, _ []atls.Validator, _ *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error) {
+				if addr == "b:7777" {
+					return newFakeClient(t), func() error { return nil }, nil
+				}
+				return nil, nil, assert.AnError
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			require := require.New(t)
+			r := &Recoverer{
+				guard:      tc.guard,
+				peerGetter: tc.peerGetter,
+				issuer:     &atls.FakeIssuer{},
+				dial:       tc.dial,
+				logger:     logger,
+			}
+			err := r.recoverOnce(ctx)
+			require.ErrorIs(err, tc.wantErr)
+		})
+	}
+}
+
+func TestRecoverFromPeer(t *testing.T) {
+	require := require.New(t)
+	ctx := t.Context()
+	logger := slog.Default()
+
+	dialCalled := false
+	expectedAddr := "127.1.2.3:7777"
+	r := &Recoverer{
+		guard:  newStaleGuard(t),
+		issuer: &atls.FakeIssuer{},
+		dial: func(_ context.Context, issuer atls.Issuer, validators []atls.Validator, _ *slog.Logger, addr string) (meshapi.MeshAPIClient, func() error, error) {
+			dialCalled = true
+			require.Equal(expectedAddr, addr)
+			require.Len(validators, 1)
+			require.IsType(&snp.Validator{}, validators[0])
+			require.NotNil(issuer)
+			return newFakeClient(t), func() error { return nil }, nil
+		},
+		logger: logger,
+	}
+
+	require.NoError(r.recoverFromPeer(ctx, nil, expectedAddr))
+	require.True(dialCalled)
+}
+
+type fakeGuard struct {
```

Comment:
This seems to be a lot of code to implement the fake logic. Why aren't we using stubs instead?

Reply:
Used stubs where possible. The guard, however, needs to call the authorizer, so we need at least some fake implementation there.